### PR TITLE
Application refactoring and demo mode

### PR DESCRIPTION
**Pull request by esune**
Changes in this pr include:
- updated OpenShift deployment configurations to support BC and NS deployments.
- updated `manage` script to support demo mode using `ngrok` tunneling to connect to the BCovrin Test ledger
- refactoring of the webapps to extract all the configuration to a JSON file loaded at runtime, that is replaced in OpenShift by a configMap

I am still working on updating the application to use the schema definition that is already on the BCovrin Test ledger, rather than publishing a new schema for every new issuer - will push updates as I make progress.

@swcurran : I did a first-cut update of the readme, it should now be fairly exhaustive - let me know what your thoughts are. The part that could be improved is the once describing how to complete the demo - for now I just referenced the existing flow document.

### CODE REVIEW

**Comment by esune** [comment]
I just realized the number of filed involved in the PR is pretty high, sorry about that: I had the openshift changes ready to go in another PR, but I ended up continuing the work as the changes are interdependent.

I can open another PR just with the openshift changes, @WadeBarnes let me know what your thoughts are.

**Comment by swcurran** [comment]
Tested out the deployment in default mode.  Issues found:

1. When going from wa-admin to pasting the invited user URL+guid, it autologged me into the issuer app with the same and email address of the wa-admin user pre-set. Either give instructions to log the person out of wa-admin before going to the issuer app, or (not sure if this is possible) - deliberately block the wa-admin used from access the issuer app.  That would (I think) force the user to the login screen. 

- related - ideal if the name and user ID from guid could be picked up if there is no OIDC token from which to pull data.

2. The QR code gave a broken image when displayed, and the string with "Streetcred URL..." was not a clickable link.  If not clickable, it should not be there. OK to remove it.

Everything else seemed to run fine.  I'm not a fan of the text of the email error message, but that should be easy to clean up.

Didn't try the local only approach.  If you have tried them all, I'm good with merging this.

File: openshift/templates/identity-kit-admin/identity-kit-admin-deploy.ns.param [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#=========================================================
+# OpenShift template parameters for:
+# Component: .
+# Template File: ../openshift/templates/identity-kit-admin/identity-kit-admin-deploy.json
+#=========================================================
+NAME=ns-id-admin
+APP_NAME=ns-id
+APP_GROUP=ns-id
+SUFFIX=
```
Comment (WadeBarnes) [inline openshift/templates/identity-kit-admin/identity-kit-admin-deploy.ns.param]: Should a suffix be defined to ensure resources are created with unique names?

File: openshift/templates/identity-kit-public/identity-kit-public-deploy.bc.param [hunk 2]
```diff
@@ -4,17 +4,23 @@
...
 CADDY_CONFIG_MOUNT_PATH=/etc/
+APPCONFIG_CONFIG_FILE_NAME=config.json
+APPCONFIG_CONFIG_MAP_NAME=identity-kit-admin-appconfig
+APPCONFIG_CONFIG_MOUNT_PATH=/srv/assets/config/
 WEB_HOST_NAME=0.0.0.0
 WEB_HOST_PORT=8080
 WEB_HOST_TLS=off
 API_HOST=identity-kit-controller
 API_PORT=8080
-APPLICATION_DOMAIN=identity-kit.pathfinder.gov.bc.ca
```
Comment (WadeBarnes) [inline openshift/templates/identity-kit-public/identity-kit-public-deploy.bc.param]: I don't think you meant to make this change.

Comment (esune) [reply openshift/templates/identity-kit-public/identity-kit-public-deploy.bc.param]: This is intended: the `prod` deployment file hasthe right host, so I kept the base configuration generic.

File: docker/manage [hunk 3]
```diff
@@ -342,18 +392,42 @@ case "${COMMAND}" in
...
     ;;
+  start-demo)
+    isJQInstalled
+    isCurlInstalled
+    isNgrokInstalled
+
+    
+    export BCOVRIN_SCHEMA_ID="85459GxjNySJ8HwTTQ4vq7:2:verified_person:1.22"
+    export LEDGER_URL="http://test.bcovrin.vonx.io"
+    export NGROK_AGENT_ENDPOINT=$(curl http://localhost:4040/api/tunnels | ${JQ_EXE} --raw-output '.tunnels | map(select(.name | contains("identity-kit-agent"))) | .[0] | .public_url')
```
[the remark this answers is not in the document]
Comment (esune) [reply docker/manage]: Good catch, it should be fixed now.

File: openshift/templates/identity-kit-admin/identity-kit-admin-deploy.ns.param (hunk 1 above)
Comment (esune) [reply openshift/templates/identity-kit-admin/identity-kit-admin-deploy.ns.param]: I think the name of the Nova Scotia instance is different enough from the base app that it is not necessary (I wouldn't really be sure what suffix to assign without making the naming weird to be honest).

**Comment by esune** [comment]
@swcurran 

1. When using the same browser for the admin and public apps, that is something expected as the user belongs to the same realm. I will try what happens when having the public app on a separate realm, however it is an unlikely scenario in a "production-like" since the inviter and invitee will be two different users using different machines/browsers.

* I am not sure I understand the use-case for not having an OIDC token: aren't users supposed to be authenticating through an OIDC-compliant identity provider?

2. Not sure what the issue was with the QR code, it seems to display fine for me. My guess is that - since we're using Google's QR code generator - sometimes the request to create the code doesn't succeed and the image doesn't load. I am looking into ways to remove that dependency and using a QR generator from within the app instead.
About the link: it is clickable, it just won't work (e.g.: it will open the browser's error page) unless you are on iOS (haven't tried Android) as it is a deeplink to the StreetCred app. This is the same we're doing in IIWBook/Email Verification. Should I still remove it?

**Comment by esune** [comment]
@swcurran 

1. When using the same browser for the admin and public apps, that is something expected as the user belongs to the same realm. I will try what happens when having the public app on a separate realm, however it is an unlikely scenario in a "production-like" since the inviter and invitee will be two different users using different machines/browsers.

* I am not sure I understand the use-case for not having an OIDC token: aren't users supposed to be authenticating through an OIDC-compliant identity provider?

2. Not sure what the issue was with the QR code, it seems to display fine for me. My guess is that - since we're using Google's QR code generator - sometimes the request to create the code doesn't succeed and the image doesn't load. I am looking into ways to remove that dependency and using a QR generator from within the app instead.
About the link: it is clickable, it just won't work (e.g.: it will open the browser's error page) unless you are on iOS (haven't tried Android) as it is a deeplink to the StreetCred app. This is the same we're doing in IIWBook/Email Verification. Should I still remove it?

**Comment by esune** [comment]
@swcurran I had actually already included a note/instructions about the admin user being able to also sign-in to the public app, and how it is recommended to open the public app in a separate browser, see [here](https://github.com/esune/identity-kit-poc/tree/refactoring#keycloak-configuration-and-users).

**Pull request merged**